includes/islandora_multi_batch.inc

```diff
@@ -806,29 +806,44 @@ class IslandoraMultiBatchObject extends IslandoraBatchObject {
                   error_log('can not open ' . $this->preprocessorParameters['zipfile'] . ' with code:'. (int)$opened);
                 }
               }
+
+              // This will handle all cases, except ZIP.
+
               else {
-                // This will handle all cases, except ZIP.
                 $fileurl = islandora_multi_importer_remote_file_get(trim($this->objectInfo['data'][$method[1]]));
                 if ($fileurl) {
                   $themime = $this->getMimetype(drupal_basename($fileurl));
-                      if ($dsid == "HOCR" && $themime == "application/octet-stream") {
-                          $themime = "text/html";
-                      }
-                      if ($dsid == "OCR" && $themime == "application/octet-stream") {
-                          $themime = "text/plain";
-                      }
+                  if ($dsid == "HOCR" && $themime == "application/octet-stream") {
+                    $themime = "text/html";
+                  }
+                  if ($dsid == "OCR" && $themime == "application/octet-stream") {
+                    $themime = "text/plain";
+                  }
+
+                  // $file = system_retrieve_file($fileurl, $tmpName, FALSE, FILE_EXISTS_RENAME);
+
+                  $file = (object) array(
+                    'uid' => 1,
+                    'uri' => $fileurl,
+                    'filemime' => $themime,
+                    'status' => 1,
+                  );
+
                   $datastreams[$dsid] = array(
                     'dsid' => $dsid,
                     'label' => "$dsid datastream",
                     'mimetype' => $themime,
                     'datastream_file' => $fileurl,
                     'filename' => drupal_basename($fileurl),
                     'control_group' => 'M',
-                  // 'file' => $file,.
+                    'file' => $file,
```

Comment:
The problem I encountered is/was in the batchProcessforIngest() function (includes/islandora_multi_batch.inc).   In my operations line 486 is the real problem.  That line reads...

  `$datastreams = $this->getDatastreams($errors, $files);`

...but if you look in the getDatastreams() function (elsewhere in the same .inc file) you will see that $files is NEVER referenced, so later on at line 505 the foreach loop does NOTHING...

  `foreach ($files as $file) {...`

Then, at line 513 we loop through the $datastreams looking for...

  ```
  $ds_uri = isset($ds['file']) ?
        $ds['file']->uri :
        $ds['datastream_file'];  
```

If $ds['file'] is never set, then in my tests this also returned NOTHING.  So the remedy was to bring back the 'file' element of the $ds structure.

It just looks like $files was introduced in the call to getDatastreams as a replacement for the $ds['file'] element, but nothing was ever done with it outside of ZIP imports.

Reply:
Ok, let me follow:

```PHP
483:    $errors = array();
484:   $files = array();
485:    $datastreams = array();
486:   $datastreams = $this->getDatastreams($errors, $files);
487:    if (!empty($errors)) {
````

$files and $errors are arrays, passed by reference empty to ::getDatastreams
```PHP
 667: protected function getDatastreams(&$errors = NULL, &$files = NULL) { 
````

The idea of $file is to keep track of FILE Objects (means your declaration is fine) of Files that need to tracked by Drupal´s file usage system and bound to the Batch Set ID. This files will then never be removed from the system by Drupal Cron Until they are released.

Adding something to $file makes sense for me IF the file is being created by IMI or only used and managed by IMI or was uploaded to Drupal because IMI requested it.

In case of files provided by external sources, imagine something in the tone of ```/home/mark/preciousimage.tiff```

If you add that image to `$file` array, it will be managed by Drupal, 

This happens here
```PHP
505:  foreach ($files as $file) {
506:      $file->status &= ~FILE_STATUS_PERMANENT; // Lets make all files temporary but trust file_usage to keep them around
507:      file_save($file);
508:     file_usage_add($file, 'islandora_batch', 'islandora_batch_object', $this->getBatchId());
    }
```
which means also, Drupal will remove it, or try to remove it on the next crop.
I feel that would be dangerous. Letting Drupal manage things that were not uploaded and created by Drupal.

But in any case, 

```PHP
513:   foreach ($datastreams as $ds) {
514:     $ds_uri = isset($ds['file']) ?
515:        $ds['file']->uri :
516:       $ds['datastream_file'];
```
$ds_uri, which is used for constructing the datastream should work out? $ds['file']->uri or $ds['datastream_file'];

Well, we can talk and discuss this today. I could be missing something!

Reply:
OK, I think I see what you're suggesting.  I'm going to build a test case and make some changes to my hook and see if I can fall back on $ds['datastream_file'] instead of relying on $ds['file'].  I'm still confused by $files, since it's never populated, but we can discuss that this afternoon.